### ISSUE TITLE
Ligatures and line justification issues

### ISSUE THREAD

**Issue by Omikhleia** [sile-typesetter/sile#1362]
And after #1358, #1359 and #1361 there comes another case of improper justification.

```
\begin[class=book, papersize=6in x 9in]{document}
\neverindent
\nofolios
\script[src=packages/lorem]

\font[family=Libertinus Serif, features="+liga"]

Lorem difficult and effects affairs of finition \lorem[words=20]

\font[family=Libertinus Serif, features="-liga"]

Lorem difficult and effects affairs of finition \lorem[words=20]

\font[family=DejaVu Serif, features="+liga"]

Lorem difficult and effects affairs of infinite findings, difficult affairs and effects of infine findings \lorem[words=20]

\font[family=DejaVu Serif, features="-liga"]

Lorem difficult and effects affairs of infinite findings, difficult affairs and effects of infine findings \lorem[words=20]

\end{document}
```

![image](https://user-images.githubusercontent.com/18075640/163379735-3945bdbc-e0dc-482b-96ca-2fa04b3f9652.png)

Seems to occur with fonts having ligatures enabled, and the ligatured text is smaller than the original text. So the test uses Libertinus, but also DejaVu, to show it"s not font-specific...

By the way, I am not sure hyphenation plays a role here, I have lines with ligatures but no hyphenation that also show the same pattern:
![image](https://user-images.githubusercontent.com/18075640/163380954-501d753c-c49e-46ff-8a0e-51bcc25ec663.png)

On the bright side: if this could be fixed too, it seems to me that this is the last source of "justification issue" I saw so far in my book, besides the above-mentioned issues...

**Comment by Omikhleia** [comment]
Digging with a shovel,
1. the hyphenation and discretionaries logic actually seems wrong...
2. the line ratio computation is clearly plain wrong, and inconsistent esp. with regards to 3.
3. the nnode ouputYourself routine is pretty wrong too in the general case.

**Comment by Omikhleia** [comment]
A possible course of action here could be (as a partial workaround) to at least align (2) with (3) so that both behave accordingly and the justification issues don't occur. As of today they are inconsistent, (3) ouputs the parent nnode if not hyphenated (so ligatures are applied there) but (2) computes the width based on all split (children) nnodes and discretionaries (so a potential ligature with a break point in between is ignored).

It would still be globally wrong in the worst case (= e.g. when a word is hyphenated and the hyphenation doesn't occur in between a ligature, as in "dif•fi•cult" if hyphenated as "diffi-cult").... Still the line would now be correctly justified, but the ligature cancelled:

![image](https://user-images.githubusercontent.com/18075640/163589105-c6509f19-b7c5-4273-b95a-4c631957de75.png)

Fixing that would require a much bigger effort and huge (it seems to me) refactor on (1) first... What (1) does wrong is that it doesn't exactly implement discretionaries as a prebreak/postbreak/nobreak triplet as in TeX, and that is just use a parent link to the whole un-hyphenated content. E.g. "difficult" is expanded to "N(dif)D(-)N(fi)D(-)N(cult)" and each nnode there points to "N(difficult)...

It seems to me the workaround is somewhat acceptable for a short-term solution:  this case seldom occurs and is less noticeable that incorrectly justified lines. I don't know... @simoncozens @alerque should I make a PR with the proposed workaround?

**Comment by Omikhleia** [comment]
Very similar to #223 from 2015 (which wasn't really closed actually).

**Comment by alerque** [comment]
Yes this looks like a duplicate of #223. Lets just keep this iteration open since it has more detailed info at this point.

**Comment by alerque** [comment]
I think the workaround we implemented for this is what broke my workaround for #355 that I've been using for years. Now I'm hunting for a real fix.... I *may* have come up with something but it is very hard to tell with *this* bug in the way. I started digging into this and got scared (and tired) and am having to run away at least for the night. Wow what a mess.

**Comment by alerque** [comment]
This isn't the *only* root problem, but while working on #355 I ran across this like which is surely wrong and at least exacerbating any other problems building up to cause this issue:

https://github.com/sile-typesetter/sile/blob/master/core/break.lua#L413

This is adding the width of all possible parts of discretionaries to the line even when they are *not* used.

**Comment by Omikhleia** [comment]
> This is adding the width of all possible parts of discretionaries to the line even when they are _not_ used.

Adding prebreak and postbreak but substracting replacement (at this point, i.e. when creating new active nodes). This does seem to be what the algorithm (hard to read, AFAIC, not even that sure I am looking at the right place) does, in Knuth's words:

> When `cur_p` is a discretionary break, the length of a line "from `cur_p` to `cur_p`'' has to be defined properly so that the other calculations work out. Suppose that the pre-break text at `cur_p` has length $l_0$, the post-break text has length $l_1$, and the replacement text has length $l$. Suppose also that `q` is the node following the replacement text. Then length of a line from `cur_p` to `q` will be computed as $\gamma+\beta(q)-\alpha(cur_p)$, where $\beta(q)=\beta(cur_p)-l_0+l$. 

> The actual length will be the background plus $l_1$, so the length from `cur_p` to `cur_p` should be $\gamma+l_0+l_1-l$.
If the post-break text of the discretionary is empty, a break may also discard `q`; in that unusual case we subtract the length of `q` and any other nodes that will be discarded after the discretionary break.

(It's hard to tell from that, but the second paragraph _seems_ to correspond to the code.)

**Comment by alerque** [comment]
> Adding prebreak and postbreak but substracting replacement (at this point, i.e. when creating new active nodes). This does seem to be what the algorithm (hard to read, AFAIC, not even that sure I am looking at the right place) does, in Knuth's words:

Yes, since ranting about it I did find another places where it backs out the calculation. Reading some TeX source (my eyes are bleeding) seems to confirm we're doing something similar. The difference is I don't think we need to be because we can hold more stuff in memory and just discard what we don't use and go back to what we stashed somewhere rather than reversing the math.

By the way I think I've fixed #355 (after 5 years of working around it) but #583 seems to be different and more closely related to this ligature issue. Not *quite* the same but similar.

**Comment by alerque** [comment]
As of now, the MWE in this issue is no longer reproducible. I am not sure whether that's because all the unrelated fixed added up to actually fix this, or whether the test just wasn't robust enough to survive incidental bugs being fixed.

**Comment by Omikhleia** [comment]
It actually never reproduced since we got #1378 in 0.13.

It was only left open at the time as it was felt the latter was kind of a workaround (and indeed a lot of `computeLineRatio()` logic and related node `outputYourself()` method was unclear... but since we went further fixing these in #1754,[^1] I think we can indeed close it now.

[^1]: I had noted in a comment above on Apr. 2022 that "the hyphenation and discretionaries logic actually seems wrong..." That was one of the reason for not closing it, as I remember = I knew the fix was partial and we were probably missing something... Eventually that's what #1754 confirmed and fixed. So we are now good here too... or at least we expect for the best :smile:
